ml/preprocessing/feature_reduction.py

```diff
@@ -1,30 +1,56 @@
 from sklearn.feature_selection import RFECV
 from sklearn.svm import SVR
-
 from configs import N_CV_FEATURE_REDUCTION
 from utils.log import log
 
 
-def perform_feature_reduction(x, y):
+def feature_selection_rfecv(estimator, X, y):
     """
-    Performs feature reduction in the x, y
-
-    For now, it uses linear SVR as estimator, and removes feature by feature.
+    Performs feature reduction on X with y labels for the given estimator with RFE-CV.
 
-    :param x: feature values
+    :param estimator: RFE trains this estimator with the features, and recursively removes them
+    :param X: feature values
     :param y: labels
-    :return: x, y, where x only contain the relevant features.
-    """
 
-    estimator = SVR(kernel="linear")
+    :return: x, where x only contains the relevant features.
+    """
     selector = RFECV(estimator, step=1, cv=N_CV_FEATURE_REDUCTION)
 
-    log("Features before reduction (total of {}): {}".format(len(x.columns.values), ', '.join(x.columns.values)))
-    selector.fit(x, y)
-    x = x[x.columns[selector.get_support(indices=True)]] # keeping the column names
+    selector.fit(X, y)
+    log(f"Feature ranking: {', '.join(str(e) for e in selector.ranking_)}", False)
+    log(f"Feature grid scores: {', '.join(str(e) for e in selector.grid_scores_)}", False)
+    return X[X.columns[selector.get_support(indices=True)]] # keeping the column names
 
-    log("Features after reduction (total of {}): {}".format(len(x.columns.values), ', '.join(x.columns.values)))
-    log("Feature ranking: {}".format(', '.join(str(e) for e in selector.ranking_)))
-    log("Feature grid scores: {}".format(', '.join(str(e) for e in selector.grid_scores_)))
 
-    return x
+def perform_feature_reduction(estimator, X, y, allowed_features=None):
+    """
+    Reduce the features of X for the estimator, or reduce the features to the given list of features.
+
+    :param :
+        estimator: RFE trains this estimator with the features, and recursively removes them
+        X: feature values
+        y: labels
+        allowed_features: reduce the features to this list
+
+    :return :
+        features: an array with the features of the instances
+        X: X, with only potentially relevant features
+    """
+    log(f"Features before reduction (total of {len(X.columns.values)}): {', '.join(X.columns.values)}", False)
+    # let's reduce the number of features in the set
+    if allowed_features is None:
+        # not all estimators expose coef_ or feature_importances, thus we handle these cases with a default estimator
+        try:
```

Comment:
can't you do an `if` before to decide which kernel to use?

Ah, maybe it doesn't matter, right? the function will throw an exception very quickly.

Reply:
Yes, that happens immediately